### ISSUE TITLE
clippy::unimplemented doesn't detect unimplemented!() with message

### ISSUE THREAD

**Issue by ExoticMatter** [rust-lang/rust-clippy#3528]
When a message is given to `unimplemented!()` (rust-lang/rust@f51771939ba5b36ec89b9f0b51c2cbda51164676), the `unimplemented` lint doesn't detect it.

```rust
#[allow(unreachable_code)]
#[warn(clippy::unimplemented)]
fn main() {
    unimplemented!("a()"); // case 1, not detected
    unimplemented!();      // case 2
}
```

    warning: `unimplemented` should not be present in production code
     --> src/main.rs:5:5
      |
    5 |     unimplemented!();      // case 2
      |     ^^^^^^^^^^^^^^^^^
      |
    note: lint level defined here
     --> src/main.rs:2:8
      |
    2 | #[warn(clippy::unimplemented)]
      |        ^^^^^^^^^^^^^^^^^^^^^
      = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unimplemented

clippy: 29bf75cd3147d4e34a5e70bd167898a8105f7e5d (2018-12-05)

**Comment by oli-obk** [comment]
We should move https://github.com/rust-lang/rust-clippy/blob/f5831523d335bc57a6371806b2ea1978942ac490/clippy_lints/src/panic_unimplemented.rs#L62 to a `PreExpansionPass` and just check for calls to macros named `unimplemented`

**Comment by phansch** [comment]
@maxencefrenette awesome! Feel free to post here or open a work-in-progress PR if you have any questions.

**Comment by maxencefrenette** [comment]
Opened a WIP PR #4186 . I need some guidance.

**Comment by ebroto** [comment]
Fixed in #5811
